ssl/quic/quic_port.c

```diff
@@ -442,19 +445,43 @@ int ossl_quic_port_set_net_wbio(QUIC_PORT *port, BIO *net_wbio)
  * ============================
  */
 
-static SSL *port_new_handshake_layer(QUIC_PORT *port)
+static SSL *port_new_handshake_layer(QUIC_PORT *port, QUIC_CHANNEL *ch)
 {
     SSL *tls = NULL;
     SSL_CONNECTION *tls_conn = NULL;
+    SSL *user_ssl = NULL;
+    QUIC_CONNECTION *qc = NULL;
+    QUIC_LISTENER *ql = NULL;
+
+    if (port->get_conn_user_ssl != NULL) {
+        user_ssl = port->get_conn_user_ssl(ch, port->user_ssl_arg);
+        if (user_ssl == NULL)
+            return NULL;
+        qc = (QUIC_CONNECTION *)user_ssl;
+        ql = (QUIC_LISTENER *)port->user_ssl_arg;
+    }
+
+    tls = ossl_ssl_connection_new_int(port->channel_ctx, user_ssl, TLS_method());
+    if (tls == NULL || (tls_conn = SSL_CONNECTION_FROM_SSL(tls)) == NULL) {
+        SSL_free(user_ssl);
+        return NULL;
+    }
 
     /*
-     * TODO(QUIC SERVER): NULL below needs to be replaced with a real user SSL
-     * object of either the listener or the domain which is associated with
-     * the port. https://github.com/openssl/project/issues/918
+     * If we got a user ssl, which will be embedded in a quic connection
+     * we need to fix up the connections tls pointer here
      */
-    tls = ossl_ssl_connection_new_int(port->channel_ctx, NULL, TLS_method());
-    if (tls == NULL || (tls_conn = SSL_CONNECTION_FROM_SSL(tls)) == NULL)
-        return NULL;
+    if (qc != NULL)
+        qc->tls = tls;
+
+    if (ql != NULL && ql->obj.ssl.ctx->new_pending_conn_cb != NULL)
+        if (!ql->obj.ssl.ctx->new_pending_conn_cb(ql->obj.ssl.ctx, user_ssl,
+                                                  ql->obj.ssl.ctx->new_pending_conn_arg)) {
+            SSL_free(tls);
+            SSL_free(user_ssl);
+            qc->tls = NULL;
```

Comment:
I think this needs test `if (qc != NULL)` to avoid a NULL pointer derference. We do the same test at line 474